Improve codegen diagnostic handling

nnethercote opened the pull request:
Clarify the workings of the temporary `Diagnostic` type used to send diagnostics from codegen threads to the main thread.

r? @estebank

nnethercote committed:
Overhaul `rustc_codegen_ssa::back::write::Diagnostic`.

- Make it more closely match `rustc_errors::Diagnostic`, by making the
  field names match, and adding `children`, which requires adding
  `rustc_codegen_ssa::back::write::Subdiagnostic`.
- Check that we aren't missing important info when converting
  diagnostics.
- Add better comments.
- Tweak `rustc_errors::Diagnostic::replace_args` so that we don't need
  to do any cloning when converting diagnostics.

nnethercote committed:
Remove `SharedEmitterMessage::AbortIfErrors`.

It's always paired wth `SharedEmitterMessage::Diagnostic`, so the two
can be merged.

nnethercote commented:
This is a rehash of what I attempted in #120575, before I learned that `rustc_errors::Diagnostic` doesn't impl `Send` in all configurations.

bors commented:
:umbrella: The latest upstream changes (presumably #120931) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote committed:
Overhaul `Diagnostic` args.

First, introduce a typedef `DiagnosticArgMap`.

Second, make the `args` field public, and remove the `args` getter and
`replace_args` setter. These were necessary previously because the getter
had a `#[allow(rustc::potential_query_instability)]` attribute, but that
was removed in #120931 when the args were changed from `FxHashMap` to
`FxIndexMap`. (All the other `Diagnostic` fields are public.)

nnethercote commented:
> ☔ The latest upstream changes (presumably #120931) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

I fixed the conflicts, and that change actually let me do an additional cleanup involving `Diagnostic::args`, which is nice.

cc @chenyukang

bors commented:
:umbrella: The latest upstream changes (presumably #120576) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased.

bors commented:
:umbrella: The latest upstream changes (presumably #121400) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
Another rebase, woo!

estebank commented:
@bors r+

bors commented:
:pushpin: Commit 6efffd723bc87f3a577f4f52fcffb6335406e9cb has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank 6efffd723bc87f3a577f4f52fcffb6335406e9cb -->
<!-- homu: {"type":"Approved","sha":"6efffd723bc87f3a577f4f52fcffb6335406e9cb","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8001216804/job/21856156107)
Approved by: estebank
Pushing f70f19fef41cfdda75c92f163434c29ad046cf09 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"estebank","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8001216804/job/21856156107"},"merge_sha":"f70f19fef41cfdda75c92f163434c29ad046cf09"} -->
